### PR TITLE
Aggregate testing

### LINKED ISSUE #2774: Implement Aggregate Test Runner in simulated_test_fixture.py
### Description of the task

Currently, we have a SimulatedTestRunner in simulated_test_fixture.py which takes in command line arguments and has methods to run a pass/fail test. We want to make this an Invariant Test Runner, and also make a new Aggregate Test Runner. There are a lot of common methods between these classes, so we can make a super class with those methods and extend it with these test runners.

Aggregate Test Runner should be able to run the same test for a list of parameters, keep track of validation for each run, and calculate overall success for these tests.

We also need to add a new function to validation.py to validate an aggregate test. This should loop through all the validation protos, but instead of throwing an error upon failure, keep track of which validations pass and fail and return it as a dictionary.

The Aggregate Test Fixture should calculate the overall success over all validation protos for each iteration.

The Pytest fixture should take the command line argument for aggregate testing and initialize the correct simulated test runner

### Acceptance criteria

- [ ] Rename Simulated Test Runner to Invariant Test Runner
- [ ] Create Aggregate Test runner class and extract common methods into a super class
- [ ] Implement Aggregate success calculation in the Aggregate Test Runner class
- [ ] Configure Pytest fixture to use the right test runner depending on command line arguments
- [ ] Modify validation.py to support aggregate validation

### PR DESCRIPTION
**Pull request by sauravbanna**
### Description

Implementation for Aggregate Testing RFC

### Testing Done

Builds Successfully and tests run normally

### Resolved Issues

Resolves #2774 

### TODO work

Still working on formatting output from aggregate tests and displaying it in the terminal

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/software/simulated_tests/simulated_test_fixture.py [hunk 1]
```diff
@@ -94,133 +100,153 @@ def time_provider(self):
...
 
-            """
-            time.sleep(delay)
+            tick = SimulatorTick(milliseconds=tick_duration_s * MILLISECONDS_PER_SECOND)
+            self.simulator_proto_unix_io.send_proto(SimulatorTick, tick)
+            time_elapsed_s += tick_duration_s
 
             if self.thunderscope:
-                self.thunderscope.close()
+                time.sleep(tick_duration_s)
```
Comment (itsarune) [inline src/software/simulated_tests/simulated_test_fixture.py:174]: what's this for?

File: src/software/simulated_tests/simulated_test_fixture.py [hunk 2]
```diff
@@ -94,133 +100,153 @@ def time_provider(self):
...
-                        ValidationProtoSet, always_validation_proto_set
-                    )
+        self.__stopper()
 
-                # Check that all always validations are always valid
-                validation.check_validation(always_validation_proto_set)
 
-            # Check that all eventually validations are eventually valid
-            validation.check_validation(eventually_validation_proto_set)
+class InvariantTestRunner(SimulatedTestRunner):
```
Comment (itsarune) [inline src/software/simulated_tests/simulated_test_fixture.py:296]: put class docs for what an InvariantTestRunner is

File: src/software/simulated_tests/simulated_test_fixture.py [hunk 3]
```diff
@@ -230,7 +256,82 @@ def excepthook(args):
...
-            __runner()
+            self.runner(
+                inv_always_validation_sequence_set,
+                inv_eventually_validation_sequence_set,
+                test_timeout_s[0] if type(test_timeout_s) == list else test_timeout_s,
+                tick_duration_s,
+            )
+
+
+class AggregateTestRunner(SimulatedTestRunner):
```
Comment (itsarune) [inline src/software/simulated_tests/simulated_test_fixture.py]: put class docs for what an aggregatetestrunner is

File: src/software/ai/hl/stp/play/ball_placement_play_test.py [hunk 4]
```diff
@@ -12,92 +12,97 @@
...
+                ball_velocity=tbots.Vector(0, 0),
+            ),
+        )
 
     simulated_test_runner.run_test(
-        eventually_validation_sequence_set=eventually_validation_sequence_set,
-        always_validation_sequence_set=always_validation_sequence_set,
-        test_timeout_s=test_duration,
+        setup=test_setup,
+        params=[False, False, False, False],
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/ball_placement_play_test.py:88]: what do these params correspond to?

File: src/software/ai/hl/stp/play/ball_placement_play_test.py [hunk 5]
```diff
@@ -12,92 +12,97 @@
...
+def test_two_ai_ball_placement(simulated_test_runner):
 
-    # Force play override here
-    blue_play = Play()
-    blue_play.name = PlayName.BallPlacementPlay
+    # placement point must be Vector2 to work with game controller
+    ball_final_pos = tbots.Point(-3, -2)
 
-    simulated_test_runner.blue_full_system_proto_unix_io.send_proto(Play, blue_play)
+    def test_setup(run_enemy_ai):
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/ball_placement_play_test.py]: let's rename this to `setup` because rn test_setup suggests that we're testing the setup

**Review by itsarune — COMMENTED** [review]
great work putting it up!

I had a question, does ball_placement_play on aggregate/invariant work on your computer

**Comment by sauravbanna** [comment]
> great work putting it up!
> 
> I had a question, does ball_placement_play on aggregate/invariant work on your computer

My bad, I forgot to include the main method in that test to call pytest_main

File: src/software/ai/hl/stp/play/ball_placement_play_test.py (hunk 5 above)
Comment (sauravbanna) [reply src/software/ai/hl/stp/play/ball_placement_play_test.py]: done

File: src/software/simulated_tests/simulated_test_fixture.py (hunk 2 above)
Comment (sauravbanna) [reply src/software/simulated_tests/simulated_test_fixture.py:296]: done

File: src/software/simulated_tests/simulated_test_fixture.py (hunk 3 above)
Comment (sauravbanna) [reply src/software/simulated_tests/simulated_test_fixture.py]: done

File: src/software/ai/hl/stp/play/ball_placement_play_test.py (hunk 4 above)
Comment (sauravbanna) [reply src/software/ai/hl/stp/play/ball_placement_play_test.py:88]: so for this test, it takes in a boolean parameter indicating whether enemy ai should be loaded or not. these params mean that the test should be run 4 times, and for all times, enemy ai should not be loaded.

File: src/software/simulated_tests/simulated_test_fixture.py (hunk 1 above)
Comment (sauravbanna) [reply src/software/simulated_tests/simulated_test_fixture.py:174]: this was in the old implementation of the test runner, so I kept it the same. I can't run thunderscope right now as I'm running into an error, but it looks like maybe thunderscope can only receive validation protos at 60Hz and not faster so this artificial delay is there

File: src/software/ai/hl/stp/play/shoot_or_chip_play_test.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+)
+from proto.message_translation.tbots_protobuf import create_world_state
+from proto.ssl_gc_common_pb2 import Team
+from proto.geometry_pb2 import Point, Vector, Angle, AngularVelocity
+
+from proto.ssl_gc_geometry_pb2 import Vector2
+
+
+def test_shoot_or_chip_play(simulated_test_runner):
+    def setup(arg):
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/shoot_or_chip_play_test.py]: why do we need `arg` here

**Review by itsarune — COMMENTED** [review]
I still can't seem to run shoot_or_chip or ball_placement_play. this is the error I get with this command `bazel test //software/ai/hl/stp/play:ball_placement_play_test --cache_test_results=false`
```
software/ai/hl/stp/play/ball_placement_play_test.py::test_two_ai_ball_placement ERROR

==================================== ERRORS ====================================
_________________ ERROR at setup of test_two_ai_ball_placement _________________

    @pytest.fixture
    def simulated_test_runner():
        args = load_command_line_arguments()
        tscope = None
    
        aggregate = args.aggregate
    
        simulator_proto_unix_io = ProtoUnixIO()
        yellow_full_system_proto_unix_io = ProtoUnixIO()
        blue_full_system_proto_unix_io = ProtoUnixIO()
    
        # Grab the current test name to store the proto log for the test case
        current_test = os.environ.get("PYTEST_CURRENT_TEST").split(":")[-1].split(" ")[0]
        current_test = current_test.replace("]", "")
        current_test = current_test.replace("[", "-")
    
        test_name = current_test.split("-")[0]
    
        # Launch all binaries
        with Simulator(
            f"{args.simulator_runtime_dir}/test/{test_name}", args.debug_simulator
        ) as simulator, FullSystem(
            f"{args.blue_full_system_runtime_dir}/test/{test_name}",
            args.debug_blue_full_system,
            False,
            should_restart_on_crash=False,
        ) as blue_fs, FullSystem(
            f"{args.yellow_full_system_runtime_dir}/test/{test_name}",
            args.debug_yellow_full_system,
            True,
            should_restart_on_crash=False,
        ) as yellow_fs:
>           with Gamecontroller(
                supress_logs=(not args.show_gamecontroller_logs), ci_mode=True,
            ) as gamecontroller:

software/simulated_tests/simulated_test_fixture.py:470: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <software.thunderscope.binary_context_managers.Gamecontroller object at 0x7f829306a9a0>

    def __enter__(self):
        """Enter the gamecontroller context manager.
    
        :return: gamecontroller context managed instance
    
        """
        command = ["/opt/tbotspython/gamecontroller"]
    
        if self.ci_mode:
            command = ["/opt/tbotspython/gamecontroller", "--timeAcquisitionMode", "ci"]
    
        command += ["-publishAddress", f"{self.REFEREE_IP}:{self.referee_port}"]
        command += ["-ciAddress", f"localhost:{self.ci_port}"]
    
        if self.supress_logs:
            with open(os.devnull, "w") as fp:
                self.gamecontroller_proc = Popen(command, stdout=fp, stderr=fp)
    
        else:
            self.gamecontroller_proc = Popen(command)
    
        if self.ci_mode:
            # We can't connect to the ci port right away, it takes
            # CI_MODE_LAUNCH_DELAY_S to start up the gamecontroller
            time.sleep(Gamecontroller.CI_MODE_LAUNCH_DELAY_S)
    
            self.ci_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
>           self.ci_socket.connect(("", self.ci_port))
E           ConnectionRefusedError: [Errno 111] Connection refused

software/thunderscope/binary_context_managers.py:402: ConnectionRefusedError
=========================== short test summary info ============================
ERROR software/ai/hl/stp/play/ball_placement_play_test.py::test_two_ai_ball_placement
```

File: src/software/ai/hl/stp/play/shoot_or_chip_play_test.py (hunk 6 above)
Comment (sauravbanna) [reply src/software/ai/hl/stp/play/shoot_or_chip_play_test.py]: so the setup function is called with each param, so if we define setup as taking in no arguments this would cause an error. I'll replace it with *args to make it more clear.

File: src/software/simulated_tests/simulated_test_fixture.py [hunk 7]
```diff
@@ -94,133 +100,157 @@ def time_provider(self):
...
+        If the test Thread will raises an exception we won't be able to close
+        the window from the main thread.
+
+        :param args: The args passed in from the hook
+
+        """
+
+        __stopper(delay=PAUSE_AFTER_FAIL_DELAY_S)
+        self.last_exception = args.exc_value
+        raise self.last_exception
```
Comment (itsarune) [inline src/software/simulated_tests/simulated_test_fixture.py]: good catch

File: src/software/ai/hl/stp/play/ball_placement_play_test.py [hunk 8]
```diff
@@ -12,90 +12,99 @@
...
+        ],
+        ag_always_validation_sequence_set=[[]],
+        ag_eventually_validation_sequence_set=[
+            [
+                # Ball should arrive within 5cm of placement point
+                BallEventuallyEntersRegion(
+                    regions=[tbots.Circle(ball_final_pos, 0.05)]
+                ),
+            ]
+        ],
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/ball_placement_play_test.py:106]: instead of duplicating validation for ag and inv, what do you think of getting aggregate tests also running invariant testing (since in theory, invariants should be less strict than the aggregate criteria)

Comment (itsarune) [reply src/software/ai/hl/stp/play/ball_placement_play_test.py:106]: as talked about on saturday, we'll address this in the next PR

File: src/software/simulated_tests/simulated_test_fixture.py [hunk 9]
```diff
@@ -230,7 +260,90 @@ def excepthook(args):
...
+                        if type(test_timeout_s) == list
+                        else test_timeout_s,
+                        tick_duration_s,
+                    )
+            except AssertionError:
+                failed_tests += 1
+
+        print(f"{failed_tests} test failed")
+
+        assert failed_tests == 0
```
Comment (itsarune) [inline src/software/simulated_tests/simulated_test_fixture.py:405]: the pass/fail rate percentage will be addressed in another PR, correct?

**Review by itsarune — COMMENTED** [review]
I ran `./tbots.py test shoot_or_chip_play_test --aggregate -t` and saw nothing on the visualizer

I also saw this warning in the error log, which might help:
`=============================== warnings summary ===============================
software/simulated_tests/simulated_test_fixture.py:40
  /home/arun/.cache/bazel/_bazel_arun/2ceaa17a6a488c4af80f2dad2aa4e7fe/sandbox/linux-sandbox/365/execroot/__main__/bazel-out/k8-fastbuild/bin/software/ai/hl/stp/play/shoot_or_chip_play_test.runfiles/__main__/software/simulated_tests/simulated_test_fixture.py:40: PytestCollectionWarning: cannot collect test class 'TestType' because it has a __new__ constructor (from: software/ai/hl/stp/play/shoot_or_chip_play_test.py)
    class TestType(Enum):

-- Docs: https://docs.pytest.org/en/stable/warnings.html
`

File: src/software/simulated_tests/simulated_test_fixture.py [hunk 10]
```diff
@@ -94,133 +100,157 @@ def time_provider(self):
...
 
-        def excepthook(args):
-            """This function is _critical_ for show_thunderscope to work.
-            If the test Thread will raises an exception we won't be able to close
-            the window from the main thread.
+    Test passes or fails based on the outcome of this test
+    """
 
-            :param args: The args passed in from the hook
+    def __int__(self, *args, **kwargs):
```
Comment (itsarune) [inline src/software/simulated_tests/simulated_test_fixture.py]: is this supposed to be `__init__`?

File: src/software/simulated_tests/simulated_test_fixture.py [hunk 11]
```diff
@@ -36,7 +37,12 @@
 PAUSE_AFTER_FAIL_DELAY_S = 3
 
 
-class SimulatorTestRunner(object):
+class TestType(Enum):
+    INVARIANT = 1
+    AGGREGATE = 1
+
```
Comment (itsarune) [inline src/software/simulated_tests/simulated_test_fixture.py]: let's have docs for this as well

File: src/software/simulated_tests/simulated_test_fixture.py (hunk 9 above)
Comment (sauravbanna) [reply src/software/simulated_tests/simulated_test_fixture.py:405]: Yes once we work on the validation

File: src/software/thunderscope/thunderscope.py [hunk 12]
```diff
@@ -121,7 +121,7 @@ def __init__(
         self.app = pyqtgraph.mkQApp("Thunderscope")
 
         # Setup stylesheet
-        apply_stylesheet(self.app, theme="dark_blue.xml")
+        # apply_stylesheet(self.app, theme="dark_blue.xml")
```
Comment (itsarune) [inline src/software/thunderscope/thunderscope.py]: this is an accidental commit

File: src/software/simulated_tests/simulated_test_fixture.py [hunk 13]
```diff
@@ -94,133 +107,157 @@ def time_provider(self):
...
         # If thunderscope is enabled, run the test in a thread and show
         # thunderscope on this thread. The excepthook is setup to catch
         # any test failures and propagate them to the main thread
         if self.thunderscope:
 
-            run_sim_thread = threading.Thread(target=__runner, daemon=True)
+            run_sim_thread = threading.Thread(target=self.runner, daemon=True)
             run_sim_thread.start()
             self.thunderscope.show()
             run_sim_thread.join()
```
Comment (itsarune) [inline src/software/simulated_tests/simulated_test_fixture.py]: in thunderscope mode, it looks like we ignore the actual test timeouts and timeout in 3 seconds (default)

**Review by itsarune — COMMENTED** [review]
few more things, almost there 👌

File: src/software/simulated_tests/simulated_test_fixture.py (hunk 13 above)
Comment (sauravbanna) [reply src/software/simulated_tests/simulated_test_fixture.py]: yeah for some reason no args were passed before, added those

File: src/software/ai/hl/stp/play/BUILD [hunk 14]
```diff
@@ -555,7 +555,7 @@ py_test(
 )
 
 cc_test(
-    name = "shoot_or_chip_play_test",
+    name = "shoot_or_chip_play_test_old",
```
Comment (nimazareian) [inline src/software/ai/hl/stp/play/BUILD]: nit: Just to stay consistent
```suggestion
    name = "shoot_or_chip_play_cpp_test",
```

File: src/software/simulated_tests/simulated_test_fixture.py (hunk 9 above)
Comment (nimazareian) [reply src/software/simulated_tests/simulated_test_fixture.py:405]: Can we have a ticket for this (or for the aggregate validations?) so it's tracked?

File: src/software/simulated_tests/simulated_test_fixture.py [hunk 15]
```diff
@@ -255,7 +310,101 @@ def excepthook(args):
...
+                # If thunderscope is disabled, just run the test
+                else:
+                    self.runner(
+                        ag_always_validation_sequence_set,
+                        ag_eventually_validation_sequence_set,
+                        test_timeout_s[x]
+                        if type(test_timeout_s) == list
+                        else test_timeout_s,
+                        tick_duration_s,
+                    )
```
Comment (nimazareian) [inline src/software/simulated_tests/simulated_test_fixture.py]: This seems almost identical to the code in `InvariantTestRunner`. Can we refactor it to the parent class?

File: src/software/simulated_tests/simulated_test_fixture.py [hunk 16]
```diff
@@ -114,138 +152,155 @@ def run_test(
...
-            run_sim_thread = threading.Thread(target=__runner, daemon=True)
+            run_sim_thread = threading.Thread(
+                target=self.runner,
+                daemon=True,
+                args=[
+                    inv_always_validation_sequence_set,
+                    inv_eventually_validation_sequence_set,
+                    test_timeout_s[0]
+                    if type(test_timeout_s) == list
+                    else test_timeout_s,
```
Comment (nimazareian) [inline src/software/simulated_tests/simulated_test_fixture.py]: nit: can this be made into a variable for better readability of the args (same with `AggregateTestRunner`)

File: src/software/simulated_tests/simulated_test_fixture.py [hunk 17]
```diff
@@ -36,7 +37,19 @@
...
+class TestType(Enum):
+    """
+    Enum for the 2 types of tests available
+    """
+
+    INVARIANT = 1
+    AGGREGATE = 2
+
+    # to make pytest ignore this class
+    __test__ = False
```
Comment (nimazareian) [inline src/software/simulated_tests/simulated_test_fixture.py]: This doesn't seem to be used

File: src/software/ai/hl/stp/play/shoot_or_chip_play_test.py [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+from proto.ssl_gc_common_pb2 import Team
+from proto.geometry_pb2 import Point, Vector, Angle, AngularVelocity
+
+from proto.ssl_gc_geometry_pb2 import Vector2
+
+
+def test_shoot_or_chip_play(simulated_test_runner):
+    def setup(*args):
+        ball_initial_pos = tbots.Point(-1.4, 2)
+        tbots.Vector(0, 0)
```
Comment (nimazareian) [inline src/software/ai/hl/stp/play/shoot_or_chip_play_test.py]: This doesn't seem to be used

File: src/software/ai/hl/stp/play/shoot_or_chip_play_test.py [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+
+        simulated_test_runner.blue_full_system_proto_unix_io.send_proto(Play, blue_play)
+
+        simulated_test_runner.simulator_proto_unix_io.send_proto(
+            WorldState, world_state
+        )
+
+    simulated_test_runner.run_test(
+        setup=setup,
+        params=[0, 1, 2, 3, 4],
```
Comment (nimazareian) [inline src/software/ai/hl/stp/play/shoot_or_chip_play_test.py:80]: The args don't seem to be used. Is the point just so the test runs 5 times? If so, can you add a comment saying that it's just so the test runs multiple times, but the state is unchanged?

**Review by nimazareian — COMMENTED** [review]
Either in this PR, or as a part of a different ticket, we should refactor `simulated_test_fixture.py`. Right now it's pretty difficult to read since there are like 4 classes in here and the file is 600+ lines. IMO it'd be easier to follow to have the general test setup and args in this file, and the Test Runner implementations be spread out in their own files. @itsarune @PhilD71 @sauravbanna Curious what you guys think?

I'm also curious if there's a simple way to know if a test is aggregated, invariant, or a combination? Since I noticed that all tests fail if `--aggregate` arg is passed to a non aggregate test. Though this isn't that big of an issue.

Phil reply (for some reason I can't reply regularly): Lets hold off until we get Amr's changes in - I'm planning on keeping some of his hierarchy which should clean it up a little bit

File: src/software/ai/hl/stp/play/shoot_or_chip_play_test.py (hunk 19 above)
Comment (sauravbanna) [reply src/software/ai/hl/stp/play/shoot_or_chip_play_test.py:80]: yeah pretty much just to run the test 5 times, but if we wanted to customize each run we could pass actual values through this array. I'll add the comment

**Comment by sauravbanna** [comment]
> I'm also curious if there's a simple way to know if a test is aggregated, invariant, or a combination? Since I noticed that all tests fail if `--aggregate` arg is passed to a non aggregate test. Though this isn't that big of an issue.

So with this implementation there isn't a thing as a test being only aggregate or only invariant. right now for example in shoot_or_chip_play_test.py, we have a list of 5 values for the params. If we run in aggregate, it runs 5 times, but if we run invariant, it only runs once with the first param.

If a test doesn't ever need to be run in aggregate mode, we just need to specify a list of just 1 value for the params. That way, no matter which mode (invariant vs aggregate) its run in, the test will only run once. So I don't think we should have the issue of the '--aggregate' argument causing tests to fail once we port the tests over to use this runner

File: src/software/thunderscope/thunderscope.py (hunk 12 above)
Comment (nimazareian) [reply src/software/thunderscope/thunderscope.py]: bump

**Review by nimazareian — COMMENTED** [review]
looks good. just an outstanding comment regarding the theme

File: src/software/ai/hl/stp/play/corner_kick_play_test.py [hunk 20]
```diff
@@ -67,90 +68,63 @@ def test_corner_kick_play_bottom_left(simulated_test_runner):
...
+def test_corner_kick_play_bottom_left(simulated_test_runner):
+    # TODO- #2781 Validation
     # TODO- #2781 Validation
-    eventually_validation_sequence_set = [[]]
-
     simulated_test_runner.run_test(
-        eventually_validation_sequence_set=eventually_validation_sequence_set,
-        always_validation_sequence_set=always_validation_sequence_set,
+        setup=lambda ball_and_bots_pos: corner_kick_play_test_setup(
+            ball_and_bots_pos, simulated_test_runner
```
Comment (nimazareian) [inline src/software/ai/hl/stp/play/corner_kick_play_test.py]: nit: instead of passing an object with all of the parameters, it might make it slightly more readable to pass in individual fields of the parameter object directly, since the object is created here as well.

File: src/software/ai/hl/stp/play/free_kick_play_test.py [hunk 21]
```diff
@@ -134,125 +151,77 @@ def test_free_kick_play_enemy(simulated_test_runner, ball_pos, yellow_bot_pos):
...
+        ],
+        inv_always_validation_sequence_set=[[]],
+        inv_eventually_validation_sequence_set=[[]],
+        ag_always_validation_sequence_set=[[]],
+        ag_eventually_validation_sequence_set=[[]],
         test_timeout_s=15,
     )
 
 
+free_kick_play_test_setup.__test__ = False
```
Comment (nimazareian) [inline src/software/ai/hl/stp/play/free_kick_play_test.py]: What does this line do? Can you add a comment briefly explaining it (I think it's in other test files as well)

File: src/software/ai/hl/stp/play/defense_play_test.py [hunk 22]
```diff
@@ -32,60 +32,60 @@
     ],
 )
 def test_defense_play(simulated_test_runner, yellow_positions):
```
Comment (nimazareian) [inline src/software/ai/hl/stp/play/defense_play_test.py:34]: @itsarune The behaviour of this play seems a bit weird. One of the friendly robots is running shadow enemy and and almost blocking our attacker, and the attacker also chips the ball into the net which I don't think is allowed
[defense_play_test_chipping_and_positioning.zip](https://github.com/UBC-Thunderbots/Software/files/11069494/defense_play_test_chipping_and_positioning.zip)

File: src/software/ai/hl/stp/play/corner_kick_play_test.py [hunk 23]
```diff
@@ -67,90 +68,63 @@ def test_corner_kick_play_bottom_left(simulated_test_runner):
         ),
     )
 
-    # Always Validation
-    # TODO- #2781 Validation
-    always_validation_sequence_set = [[]]
 
-    # Eventually Validation
+def test_corner_kick_play_bottom_left(simulated_test_runner):
+    # TODO- #2781 Validation
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/corner_kick_play_test.py:65]: duplicated comment

File: src/software/ai/hl/stp/play/corner_kick_play_test.py [hunk 24]
```diff
@@ -11,20 +11,21 @@
 from proto.ssl_gc_common_pb2 import Team
 
 
-def test_corner_kick_play_bottom_left(simulated_test_runner):
-
+def corner_kick_play_test_setup(ball_and_bots_pos, simulated_test_runner):
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/corner_kick_play_test.py]: pytest is going to try to test everything with "test" in the name, so just call it corner_kick_play_setup

File: src/software/ai/hl/stp/play/free_kick_play_test.py [hunk 25]
```diff
@@ -10,34 +10,24 @@
...
 
-    yellow_bots = [
-        tbots.Point(1, 0),
-        tbots.Point(1, 2.5),
-        tbots.Point(1, -2.5),
-        tbots.Field.createSSLDivisionBField().enemyGoalCenter(),
-        tbots.Field.createSSLDivisionBField().enemyDefenseArea().negXNegYCorner(),
-        tbots.Field.createSSLDivisionBField().enemyDefenseArea().negXPosYCorner(),
-    ]
+def free_kick_play_test_setup(test_setup_arg, simulated_test_runner):
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/free_kick_play_test.py]: omit the "test" in the function name

File: src/software/simulated_tests/simulated_test_fixture.py (hunk 9 above)
Comment (itsarune) [reply src/software/simulated_tests/simulated_test_fixture.py:405]: bump

File: src/software/ai/hl/stp/play/defense_play_test.py (hunk 22 above)
Comment (itsarune) [reply src/software/ai/hl/stp/play/defense_play_test.py:34]: that's pretty funny, but DefensePlay is currently unused. Eventually william's changes will replace this

File: src/software/simulated_tests/simulated_test_fixture.py [hunk 26]
```diff
@@ -114,148 +139,260 @@ def run_test(
...
+                    ag_eventually_validation_sequence_set,
+                    test_timeout_s,
+                    tick_duration_s,
+                )
+            except AssertionError:
+                failed_tests += 1
+
+        # TODO (#2856) Fix validation and results output
+
+        print(f"{failed_tests} test failed")
```
Comment (itsarune) [inline src/software/simulated_tests/simulated_test_fixture.py]: the "failed_tests" variable is out-of-scope here, also use a logger.info statement here rather than a python print

**Review by itsarune — COMMENTED** [review]
some more stuff from me

File: src/software/simulated_tests/simulated_test_fixture.py (hunk 9 above)
Comment (sauravbanna) [reply src/software/simulated_tests/simulated_test_fixture.py:395]: sorry made the issue but forgot to add a todo

Comment (sauravbanna) [reply src/software/simulated_tests/simulated_test_fixture.py:405]: can this be a seperate PR for fixing the output? I added a TODO for that issue

**Review by itsarune — DISMISSED** [review]
lgtm :fire:

**Review by nimazareian — DISMISSED** [review]

**Review by nimazareian — DISMISSED** [review]

**Review by itsarune — DISMISSED** [review]
nice work putting all this together!

**Review by nimazareian — APPROVED** [review]

**Review by itsarune — APPROVED** [review]
